client.go

```diff
@@ -211,6 +264,242 @@ func (c *Client) monitorChannel(ctx context.Context) {
+	if err != nil {
+		return nil, nil, err
+	}
+	sechan, err := uasc.NewSecureChannel(c.endpointURL, conn, c.cfg)
+	if err != nil {
+		_ = conn.Close()
+		return nil, conn, err
+	}
+	return sechan, conn, nil
+}
+
+// repairSession repair the current session after being disconnected
+func (c *Client) repairSession() error {
+
+	s := c.Session()
+
+	if _, err := c.DetachSession(); err != nil {
+		return err
+	}
+
+	debug.Printf("Trying to reactivate existing session")
+
+	err := c.ActivateSession(s)
+	if err != nil {
+		return errors.Errorf("Session reactivation failed")
+	}
+	debug.Printf("Existing session reactivated trying to repair subscriptions")
+	if err := c.repairSubscriptions(); err != nil {
+		return err
+	}
+	debug.Printf("Subscriptions repaired")
+	return nil
+}
+
+// SubscriptionIDs gets a list of subscriptionIDs
+func (c *Client) SubscriptionIDs() []uint32 {
+	subscriptionIDs := []uint32{}
+	c.subMux.Lock()
+	defer c.subMux.Unlock()
+	for key := range c.subscriptions {
+		subscriptionIDs = append(subscriptionIDs, key)
+	}
+	return subscriptionIDs
+}
+
+// repairSubscriptions repair all the subscriptions of subscriptionIDs given,
+// if no subscriptionIDs repair all subscriptions
+func (c *Client) repairSubscriptions(subscriptionIDs ...uint32) error {
+
+	if subscriptionIDs == nil {
+		subscriptionIDs = c.SubscriptionIDs()
+	}
+
+	c.subMux.RLock()
+	defer c.subMux.RUnlock()
+
+	for _, subID := range subscriptionIDs {
+		sub, ok := c.subscriptions[subID]
+		if !ok {
+			return errors.Errorf("Invalid SubscriptionID")
+		}
+		if err := c.repairSubscription(sub); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) repairSubscription(sub *Subscription) error {
+	debug.Printf("RepairSubscription  for SubscriptionId %d", sub.SubscriptionID)
+	if status, err := c.subscriptionRepublish(sub); err != nil {
+		switch status {
+		case ua.StatusBadSessionIDInvalid:
+			return err
+		case ua.StatusBadSubscriptionIDInvalid:
+			debug.Printf("Republish failed, subscriptionId is not valid anymore on server side.")
+			// return sub.recreateSubscriptionAndMonitoredItem()
+			return errors.Errorf("Republish failed, subscriptionId is not valid anymore on server side.")
+		}
+	}
+	return nil
+}
+
+// subscriptionRepublish send republish request for the given subscription
+// republish should end with a StatusCode BadMessageNotAvailable
+// wich implies that there's no more message to restore
+func (c *Client) subscriptionRepublish(sub *Subscription) (ua.StatusCode, error) {
+	for {
+		req := &ua.RepublishRequest{
+			SubscriptionID:           sub.SubscriptionID,
+			RetransmitSequenceNumber: sub.lastSequenceNumber + 1,
+		}
+
+		debug.Printf("Republish Request for subscription %d retransmitSequenceNumber=%d",
+			req.SubscriptionID,
+			req.RetransmitSequenceNumber,
+		)
+
+		if c.sessionClosed() {
+			debug.Printf("Republish aborted")
+			return ua.StatusOK, nil
+		}
+
+		res, err := sub.republish(req)
```

Comment:
If `sub.republish()` is simplified as I suggested, then `res` and `err` cannot be simultaneously not nil here, so the following error handling can be simplified as well.

Reply:
Different than suggestion, need further discusion
Changed in e0f222ba1a18d884479a5f5c433138578982c039